test/unit/logic/peers.js

```diff
@@ -122,80 +123,97 @@ describe('peers', function () {
 			var inserted = list[0];
 			expect(list.length).equal(1);
-			expect(arePeersEqual(inserted, randomPeer)).to.be.ok;
+			expect(arePeersEqual(inserted, validPeer)).to.be.ok;
 
-			var modifiedPeer = _.clone(randomPeer);
+			var modifiedPeer = _.clone(validPeer);
 			modifiedPeer.height += 1;
 			peers.upsert(modifiedPeer);
 			list = peers.list();
 			var updated = list[0];
 			expect(list.length).equal(1);
 			expect(arePeersEqual(updated, modifiedPeer)).to.be.ok;
-			expect(arePeersEqual(updated, randomPeer)).to.be.not.ok;
+			expect(arePeersEqual(updated, validPeer)).to.be.not.ok;
 
 		});
 
 		it('should not update height with insertOnly param', function () {
-			peers.upsert(randomPeer);
+			peers.upsert(validPeer);
 			var list = peers.list();
 			var inserted = list[0];
 			expect(list.length).equal(1);
-			expect(arePeersEqual(inserted, randomPeer)).to.be.ok;
+			expect(arePeersEqual(inserted, validPeer)).to.be.ok;
 
-			var modifiedPeer = _.clone(randomPeer);
+			var modifiedPeer = _.clone(validPeer);
 			modifiedPeer.height += 1;
 			peers.upsert(modifiedPeer, true);
 			list = peers.list();
 			var updated = list[0];
 			expect(list.length).equal(1);
 			expect(arePeersEqual(updated, modifiedPeer)).to.be.not.ok;
-			expect(arePeersEqual(updated, randomPeer)).to.be.ok;
+			expect(arePeersEqual(updated, validPeer)).to.be.ok;
 		});
 
 		it('should insert peer with different ports', function () {
-			peers.upsert(randomPeer);
+			peers.upsert(validPeer);
 			expect(peers.list().length).equal(1);
 
-			var differentPortPeer = _.clone(randomPeer);
+			var differentPortPeer = _.clone(validPeer);
 			differentPortPeer.nonce = 'differentNonce';
 			differentPortPeer.port += 1;
 			peers.upsert(differentPortPeer);
 			var list = peers.list();
 			expect(list.length).equal(2);
 
-			var demandedPorts = _.map([randomPeer, differentPortPeer], 'port');
+			var demandedPorts = _.map([validPeer, differentPortPeer], 'port');
 			var listPorts = _.map(list, 'port');
 
 			expect(_.isEqual(demandedPorts.sort(), listPorts.sort())).to.be.ok;
 		});
 
 		it('should insert peer with different ips', function () {
-			peers.upsert(randomPeer);
+			peers.upsert(validPeer);
 			expect(peers.list().length).equal(1);
 
-			var differentIpPeer = _.clone(randomPeer);
+			var differentIpPeer = _.clone(validPeer);
+			delete differentIpPeer.string;
 			differentIpPeer.ip = '40.40.40.41';
 			differentIpPeer.nonce = 'differentNonce';
 
-			expect(differentIpPeer.ip).to.not.equal(randomPeer);
-			peers.upsert(differentIpPeer);
+			console.log(peers.upsert(differentIpPeer));
 			var list = peers.list();
 			expect(list.length).equal(2);
 
-			var demandedIps = _.map([randomPeer, differentIpPeer], 'ip');
+			var demandedIps = _.map([validPeer, differentIpPeer], 'ip');
 			var listIps = _.map(list, 'ip');
 
 			expect(_.isEqual(demandedIps.sort(), listIps.sort())).to.be.ok;
 		});
+
+		describe('should fail with valid error code', function () {
+
```

Comment:
here only `ON_MASTER` error codes can be covered, I've added a tests for ALREADY_ADDED , ALREADY_REMOVED, DIFFERENT_CONN_ID in rules tests. 
INVALID_HEADERS, INCOMPATIBLE_NONCE, INCOMPATIBLE_VERSION are being checked in handshake tests